semimpl/do.go

```diff
@@ -0,0 +1,73 @@
+package semimpl
+
+import (
+	"context"
+	"errors"
+	"log"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// Do concurrently executes callback at most maxIter times or until ctx is done
+// or canceled. Concurrency is handled leveraging the semaphore pattern, which
+// ensures at most numWorkers goroutines are spawned at se same time.
+func Do(ctx context.Context, numWorkers, maxIter int, callback func()) {
+	numWorkers = sanitizeNumWorkers(numWorkers)
+	maxIter = sanitizeMaxIter(maxIter)
+	callback = sanitizeCallback(callback)
+
+	sem := semaphore.NewWeighted(int64(numWorkers))
+	wg := sync.WaitGroup{}
+
+	for i := 0; i < maxIter || maxIter == 0; i++ {
+		wg.Add(1)
+
+		if err := sem.Acquire(ctx, 1); err != nil {
+			handleContextError(err)
+			wg.Done()
+			break
+		}
+
+		go func() {
+			defer func() {
+				sem.Release(1)
+				wg.Done()
+			}()
+			callback()
+		}()
+	}
+
+	wg.Wait()
+}
+
+func sanitizeNumWorkers(numWorkers int) int {
+	if numWorkers < 1 {
+		return 1
+	}
+	return numWorkers
+}
+
+func sanitizeMaxIter(maxIter int) int {
+	if maxIter < 0 {
+		return 0
+	}
+	return maxIter
+}
+
+func sanitizeCallback(callback func()) func() {
+	if callback == nil {
+		return func() {}
+	}
+	return callback
+}
```

Comment:
[suggestion] 
Instead of correcting the values provided, maybe we could raise an error and inform the user?
It is not 100% necessary though since we make a summary of the values which have been actually used in the output, the user will know if it is not what they intended to input.

Reply:
Or maybe just add a message in the output next to the value changed? ("Concurrent requests number  must be >= 1" for instance)

Reply:
If we consider the fact the results are going to be uploaded, it's indeed a good idea to stop the process for any weird value.

Which values should be weird?
- `callback == nil`, because it would iterate doing literally nothing
- `numWorkers == 0`? 0 is not logical because we can't have 0 worker, but here we allow it to mean "no concurrency". We also allow `-1` because it's quite common, and at this point why not any negative number? (someone using -732 is likely not taking it seriously)
- `maxIter == 0`? Same

`-1` is often used for this purpose, but `0` is kind of convenient here because it's Go's default value for ints so it seems quite logical that when `-c` is not set, its default value matches "no concurrency" (same with `-r` / `maxIter`)

Reply:
Alternatively, we can use a strict logic for `semimpl` (`numWorkers > 0` and `maxIter == -1 || maxIter > 0`) and handle the defaulting at a high level, in `cmd` or `request`.

Reply:
I would go for the 2nd option (`numWorkers > 0` `and maxIter > 0 || maxIter == -1`) and handle defaulting at a higher level if others agree.

I don't know for a safeguard, because I have no idea what would be a reasonable number and I think it would be the user responsibility to know roughly how much they can ask from their machine.
Or an option would be to add an arbitrary limit, but I feel like it should still be possible to explicitly override it.

Reply:
I realize there's more to do in terms of error handling: in addition to `semimpl.Do`, we must also handle invalid url or failed requests in `requesst.Do`. For that reason I suggest we create a dedicated issue for global error handling in the app.
It's not a huge piece of work, but it'll keep this PR well scoped and without behavioural changes.

Reply:
> I realize there's more to do in terms of error handling: in addition to semimpl.Do, we must also handle invalid url or failed requests in requesst.Do. For that reason I suggest we create a dedicated issue for global error handling in the app.

I don't think it should. The function it is passed must handle errors itself. `semimpl.Do` ùust not know what behavior it is wrapping. Something like: 

```go
semimpl.Do(ctx, concurrency, requests, func() {
	rec, err := doRequest(url, timeout)
	if err != nil {
		fails += 1
		return
	} else {
		records.Append(rec)
	}
})
```

Anyway, it's outside the scope of this PR :)

Reply:
Last comment: obviously, that's exactly what I meant by _handle invalid url or failed requests **IN** `request.Do`_ and by _global error handling in the app_.

> `callback == nil` should be a compile time error (although `func()` accepts `nil`)

Should we rewrite the Go compiler then? 😛 

So in the end it seems we all agree on the error handling specs, time to write the issue 🚀